crates/forge_services/src/agent_loader.rs

```diff
@@ -0,0 +1,341 @@
+            http: HttpConfig::default(),
+            stdout_max_prefix_length: 50,
+            stdout_max_suffix_length: 50,
+            stdout_max_line_length: 200,
+            max_read_size: 2000,
+            max_file_size: 1024 * 1024,
+        };
+
+        let agent_dir = environment.agent_path();
+
+        let agent_content = r#"---
+id: "test"
+title: "Test Agent"
+description: "A test agent"
+system_prompt: "Test instructions"
+---
+
+# Test Agent
+
+This is the content of the test agent.
+"#;
+
+        let mock_infra = Arc::new(
+            MockInfra::new(environment)
+                .with_file(agent_dir.join("test.md"), agent_content.to_string()),
+        );
+
+        let service = super::AgentLoaderService::new(mock_infra);
+        let actual = service.load_agents().await.unwrap();
+
+        assert_eq!(actual.len(), 1);
+        let agent = &actual[0];
+        assert_eq!(agent.id.as_str(), "test");
+        assert_eq!(agent.title.as_ref().unwrap(), "Test Agent");
+        assert_eq!(agent.description.as_ref().unwrap(), "A test agent");
+    }
+
+    #[tokio::test]
+    async fn test_load_agents_caches_results() {
+        let temp_dir = tempdir().unwrap();
+        let agent_dir = temp_dir.path().join("agents");
+
+        let environment = Environment {
+            os: "test".to_string(),
+            pid: 12345,
+            cwd: temp_dir.path().to_path_buf(),
+            home: Some(temp_dir.path().to_path_buf()),
+            shell: "bash".to_string(),
+            base_path: temp_dir.path().to_path_buf(), /* Use temp_dir as base_path so
+                                                       * agent_path() works */
+            retry_config: RetryConfig::default(),
+            max_search_lines: 25,
+            max_search_result_bytes: 256000,
+            fetch_truncation_limit: 0,
+            forge_api_url: Url::parse("http://localhost:8000").unwrap(),
+            http: HttpConfig::default(),
+            stdout_max_prefix_length: 50,
+            stdout_max_suffix_length: 50,
+            stdout_max_line_length: 200,
+            max_read_size: 2000,
+            max_file_size: 1024 * 1024,
+        };
+
+        let mock_infra = Arc::new(
+            MockInfra::new(environment).with_file(
+                agent_dir.join("test.md"),
+                r#"---
+id: "test"
+title: "Test Agent"
+description: "A test agent"
+system_prompt: "Test instructions"
+---"#
+                    .to_string(),
+            ),
+        );
+
+        let service = super::AgentLoaderService::new(mock_infra);
+
+        // First call should load from infra
+        let first_result = service.load_agents().await.unwrap();
+
+        // Second call should return cached result
+        let second_result = service.load_agents().await.unwrap();
+
+        assert_eq!(first_result.len(), second_result.len());
+        assert_eq!(first_result[0].title, second_result[0].title);
+    }
+}
```

Comment:
too much of boilerplate to test the parsing function. Better approach:
- Create .md files that contain agent details
- Move the `parse_agent_file` function out of the service and using include_str! import them and run the test.